nodes/basic-pow/src/service.rs

```diff
@@ -0,0 +1,176 @@
+	($config:expr) => {{
+		let mut import_setup = None;
+		let inherent_data_providers = crate::service::build_inherent_data_providers()?;
+
+		let builder = sc_service::ServiceBuilder::new_full::<
+			runtime::opaque::Block, runtime::RuntimeApi, crate::service::Executor
+		>($config)?
+			.with_select_chain(|_config, backend| {
+				Ok(sc_client::LongestChain::new(backend.clone()))
+			})?
+			.with_transaction_pool(|config, client, _fetcher| {
+				let pool_api = sc_transaction_pool::FullChainApi::new(client.clone());
+				Ok(sc_transaction_pool::BasicPool::new(config, std::sync::Arc::new(pool_api)))
+			})?
+			.with_import_queue(|_config, client, select_chain, _transaction_pool| {
+
+				let pow_block_import = sc_consensus_pow::PowBlockImport::new(
+					client.clone(),
+					client.clone(),
+					crate::pow::Sha3Algorithm,
+					0, // check inherents starting at block 0
+					select_chain,
+					inherent_data_providers.clone(),
+				);
+
+				let import_queue = sc_consensus_pow::import_queue(
+					Box::new(pow_block_import.clone()),
+					crate::pow::Sha3Algorithm,
+					inherent_data_providers.clone(),
+				)?;
+
+				import_setup = Some(pow_block_import);
+
+				Ok(import_queue)
+			})?;
+
+		(builder, import_setup, inherent_data_providers)
+	}}
+}
+
+/// Builds a new service for a full client.
+pub fn new_full(config: Configuration<GenesisConfig>)
+	-> Result<impl AbstractService, ServiceError>
+{
+	let is_authority = config.roles.is_authority();
+
+	// sentry nodes announce themselves as authorities to the network
+	// and should run the same protocols authorities do, but it should
+	// never actively participate in any consensus process.
+	let participates_in_consensus = is_authority && !config.sentry_mode;
+
+	let (builder, mut import_setup, inherent_data_providers) = new_full_start!(config);
+	let block_import = import_setup.take()
+		.expect("Block Import is present for Full Services or setup failed before. qed");
+
+	let service = builder
+		.with_finality_proof_provider(|_client, _backend|
+			// Question. Why bother giving () as a finality proof provider when I can just
+			// not call `with_finality_proof_provider` at all?
+			Ok(Arc::new(()) as _)
+		)?
+		.build()?;
+
+	if participates_in_consensus {
+		let proposer = sc_basic_authorship::ProposerFactory::new(
+			service.client(),
+			service.transaction_pool()
+		);
+
+		// The number of rounds of mining to try in a single call
+		let rounds = 500;
+
+		let client = service.client();
+		let select_chain = service.select_chain()
+			.ok_or(ServiceError::SelectChainRequired)?;
+
+		let can_author_with =
+			sp_consensus::CanAuthorWithNativeVersion::new(service.client().executor().clone());
+
+		sc_consensus_pow::start_mine(
+			Box::new(block_import),
+			client,
+			Sha3Algorithm,
+			proposer,
+			None, // No preruntime digests
+			rounds,
+			service.network(),
+			std::time::Duration::new(2, 0),
+			Some(select_chain), // Question: What does it mean to put None here?
```

Comment:
Putting `None` means we don't use `SelectChain` when fetching best block, but directly use `Client`'s metadata. In that case, the best block information may be outdated.